src/transport/mod.rs

```diff
@@ -1,30 +1,45 @@
-use crate::interceptor::PacketInterceptor;
-use crate::transport::tcp::TcpTransport;
-use futures::{Async, Future, Sink, Stream};
-use std::io::{Read, Write};
-use std::net::SocketAddr;
-use tokio::io;
-use tokio_io::{AsyncRead, AsyncWrite};
-use tokio_tcp::TcpStream;
+use std::{
+    io::{Read, Write},
+    net::SocketAddr,
+    sync::{
+        atomic::{AtomicU64, Ordering},
+        Arc,
+    },
+};
+
+use futures::{Async, AsyncSink, Future, Poll, Sink, StartSend, Stream};
+use slog_scope::{error, trace};
+use spsc_bip_buffer::{BipBufferReader, BipBufferWriter};
+use tokio::io::{self, AsyncRead, AsyncWrite, ReadHalf, WriteHalf};
+use tokio::net::tcp::TcpStream;
 use url::Url;
-use crate::transport::ws::WsTransport;
+
+use crate::interceptor::PacketInterceptor;
+use crate::transport::{tcp::TcpTransport, ws::WsTransport};
 
 pub mod mcs;
 pub mod tcp;
 pub mod tsrequest;
-pub mod x224;
 pub mod ws;
+pub mod x224;
 
 pub type JetFuture<T> = Box<dyn Future<Item = T, Error = io::Error> + Send>;
 pub type JetStreamType<T> = Box<dyn JetStream<Item = T, Error = io::Error> + Send>;
 pub type JetSinkType<T> = Box<dyn JetSink<SinkItem = T, SinkError = io::Error> + Send>;
 
+pub const BIP_BUFFER_LEN: usize = 8 * PART_LEN;
+const PART_LEN: usize = 16 * 1024;
+
 pub trait Transport {
     fn connect(addr: &Url) -> JetFuture<Self>
     where
         Self: Sized;
-    fn message_sink(&self) -> JetSinkType<Vec<u8>>;
-    fn message_stream(&self) -> JetStreamType<Vec<u8>>;
+    fn peer_addr(&self) -> Option<SocketAddr>;
+    fn split_transport(
```

Comment:
Because now there is no Arc Mutex, we will do tokio::split for stream and sink. After that, in theory, the transport is no longer needed - therefore, we accept self.